README.md
# react-native-manage-wallpaper

Android wallpaper manager for changing wallpaper.

<a href="https://twitter.com/intent/follow?screen_name=meharbhutta"><img src="https://img.shields.io/twitter/follow/meharbhutta.svg?style=social&logo=twitter" alt="follow on Twitter" /></a>

## NPM

- stable release version: ![version](https://img.shields.io/badge/version-1.1.2-blue.svg?cacheSeconds=2592000)
- package downloads: ![downloads](https://img.shields.io/badge/downloads-22%2Fweek-brightgreen.svg?cacheSeconds=2592000)
- [![MIT license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)

## Getting started

### Installation

- Using yarn

`$ yarn add react-native-manage-wallpaper`

- Using npm

`$ npm install react-native-manage-wallpaper --save`

### Basic Usage

#### Note: You must be using React Native 0.60.0 or higher

- Initialization of a react-native project

`$ npx react-native init AwesomeProject`

### Note: [GUIDE](https://facebook.github.io/react-native/docs/getting-started)

- Then, edit `AwesomeProject/App.js`, like this:

```javascript
import { View, TouchableOpacity, Text } from 'react-native';
import ManageWallpaper, { TYPE } from 'react-native-manage-wallpaper';

export default class App extends Component<{}> {
  _callback = res => {
    console.log('Response: ', res);
  };

  _setWallpaper = () => {
   ManageWallpaper.setWallpaper(
      {
        uri: 'https://i.pinimg.com/originals/76/5e/1d/765e1dc8cb1cc115fb3b0b39a895fdeb.jpg',
      },
      this._callback,
      TYPE.HOME,
    );
  };

  render() {
    return (
      <View
        style={{
          flex: 1,
          justifyContent: 'center',
          alignItems: 'center',
          backgroundColor: '#F5FCFF',
        }}
      >
        <TouchableOpacity
          style={{
            paddingHorizontal: 30,
            paddingVertical: 8,
            marginBottom: 24,
            borderRadius: 16,
            justifyContent: 'center',
            alignItems: 'center',
            backgroundColor: '#000000',
          }}
          onPress={this._setWallpaper}
        >
          <Text
            style={{
              fontSize: 20,
              margin: 10,
              textAlign: 'center',
              color: '#FFFFFF',
            }}
          >
            Change Home Wallpaper
          </Text>
        </TouchableOpacity>
      </View>
    );
  }
}
```

### Are you using Glide already using an AppGlideModule?

Removing RNGlideModule from `react-native-manage-wallpaper`  
If you are using Glide within your application using an AppGlideModule then you will need to prevent the inclusion of the AppGlideModule in this package.

To accomplish this you can add to `android/build.gradle`:

```
project.ext {
  excludeRNGlideModule = true
}
```

### Arguments Type

Type of arguments for `setWallpaper(source, callback, type)`

#### source

The image source (either a remote URL or a local file resource).  
The currently supported formats are `png`, `jpg`, `jpeg`, `gif`.  
e.g.  
RemoteURL: { uri: 'remote url' }  
LocalFileURL: { uri: 'file:///local_file_url' }  
Local: require('path to local file with extension')

#### callback

Callback receive a response object ({ status, msg, url }).

1. status:  
   "error" string in case of handled error  
   OR  
   "success" string in case of successful setting of wallpaper.
2. msg: description related to status
3. url: resource url

#### type

The wallpaper setting type (home, lock or both).  
e.g.  
Firstly,

```javascript
import { TYPE } from 'react-native-manage-wallpaper';
```

Then,  
For Home use TYPE.HOME  
For Lock use TYPE.LOCK  
For Home & Lock use TYPE.BOTH

##### Note: No support for setting wallpaper on Lock Screen for android version < 24 (Nougat).

### To run example

```bash
cd example
yarn or npm install
npx react-native run-android
```

#### In case of any issue follow the [GUIDE](https://facebook.github.io/react-native/docs/getting-started).

### Supported React Native Versions  

This project only aims to support the latest version of React Native.  
This simplifies the development and the testing of the project.

If you require new features or bug fixes for older versions you can fork this project.

### Credits

The idea for this modules came from liuhong1happy react-native-wallpaper-manager package. It also uses Glide, but didn't have some features I needed (home, lock, both).


### Licenses

ManageWallPaper - MIT © MeharBhutta  
Glide - BSD, part MIT and Apache 2.0. See the LICENSE file for details.


CHANGELOG.md
#### 1.1.2-5 (2019-12-05)

##### Documentation Changes

* **usage:**   source argument  updated ([ca2712ee](https://github.com/meharbhutta/react-native-manage-wallpaper/commit/ca2712eeb3114a41726290833a0a07f19df20cb1))

#### 1.1.1-4 (2019-12-04)

##### Documentation Changes

*  glide detail added ([7d0b3bdf](https://github.com/meharbhutta/react-native-manage-wallpaper/commit/7d0b3bdfcbfaa490802fc58c23ba2e0cbf1bcee3))

### 1.1.0-3 (2019-12-04)

##### Documentation Changes

*  stable release updated ([c95d1d2f](https://github.com/meharbhutta/react-native-manage-wallpaper/commit/c95d1d2f22a1379b699e861ee09eb6d2bd536e6a))

##### New Features

*  upgraded glide version ([2c529350](https://github.com/meharbhutta/react-native-manage-wallpaper/commit/2c5293506c7f5b9740c46984af9cace93b83e08c))

#### 1.0.1-2 (2019-12-04)

##### Bug Fixes

*  changelog.md added ([68f66759](https://github.com/meharbhutta/react-native-manage-wallpaper/commit/68f667599c31f8c1af7aa9375e10f8ae1435a27c))



index.js
import { NativeModules, Image, Platform } from 'react-native';

const TYPE = {
  HOME: 'home',
  LOCK: 'lock',
  BOTH: 'both',
};

const ManageWallpaper = {
  setWallpaper: (source, callback = res => {}, type) => {
    if (Platform.Version >= 24 && !type) {
      throw new Error(
        'Type is undefined\nimport {TYPE} from "react-native-manage-wallpaper\nUse TYPE.HOME"',
      );
    }

    if (!type) {
      type = TYPE.HOME;
    }

    if (typeof type !== 'string') {
      throw new Error(
        'Invalid type datatype\nimport {TYPE} from "react-native-manage-wallpaper\nUse TYPE.HOME"',
      );
    }

    if (!Object.values(TYPE).includes(type)) {
      throw new Error(
        'Invalid type\nimport {TYPE} from "react-native-manage-wallpaper\nUse TYPE.HOME"',
      );
    }

    NativeModules.ManageWallpaper.setWallpaper(Image.resolveAssetSource(source), type, callback);
  },
};

export default ManageWallpaper;

export { ManageWallpaper, TYPE };


changelog.js
#!/usr/bin/env node

const git = require('simple-git');
const generate_changelog = require('generate-changelog');
const fs = require('fs');
const idx = require('idx');
const argv = require('minimist')(process.argv.slice(1));

// eslint-disable-next-line handle-callback-err
git().tags((err, tags) => {
  const currentChangelog = fs.readFileSync('./CHANGELOG.md');
  const matched = tags && tags.latest ? tags.latest.match(/v\d+.\d+.\d+-(\d+)/) : 'v1.0.0';
  const build = (idx(matched, _ => Number(_[1])) || 0) + 1;

  generate_changelog
    .generate({
      major: argv.major,
      minor: argv.minor,
      patch: argv.patch,
    })
    .then(function(changelog) {
      const rxVersion = /\d+\.\d+\.\d+/;
      const newVersion = argv.version || idx(changelog.match(rxVersion), _ => _[0]) + `-${build}`;

      changelog = changelog.replace(rxVersion, newVersion) + currentChangelog;
      fs.writeFileSync('./CHANGELOG.md', changelog);

      const addFile = c => git().add('CHANGELOG.md', c);
      const commit = c => git().commit(`build(change-log): v${newVersion}`, c);
      const addTag = c => git().addAnnotatedTag(`v${newVersion}`, `build(tag): v${newVersion}`, c);

      addFile(() => commit(() => addTag()));
    });
});


android/src/main/java/com/mmmgbhutta/reactnative/wallpapermanager/Util.java
package com.mmmgbhutta.reactnative.wallpapermanager;

import android.os.Looper;
 
public class Util {
 
    /**
     * Throws an {@link java.lang.IllegalArgumentException} if called on a thread other than the main thread.
     */
    public static void assertMainThread() {
        if (!isOnMainThread()) {
            throw new IllegalArgumentException("You must call this method on the main thread");
        }
    }
 
    /**
     * Throws an {@link java.lang.IllegalArgumentException} if called on the main thread.
     */
    public static void assertBackgroundThread() {
        if (!isOnBackgroundThread()) {
            throw new IllegalArgumentException("You must call this method on a background thread");
        }
    }
 
    /**
     * Returns {@code true} if called on the main thread, {@code false} otherwise.
     */
    public static boolean isOnMainThread() {
        return Looper.myLooper() == Looper.getMainLooper();
    }
 
    /**
     * Returns {@code true} if called on the main thread, {@code false} otherwise.
     */
    public static boolean isOnBackgroundThread() {
        return !isOnMainThread();
    }
}


android/src/main/java/com/mmmgbhutta/reactnative/wallpapermanager/ManageWallpaperModule.java
package com.mmmgbhutta.reactnative.wallpapermanager;

import android.app.Activity;
import android.app.WallpaperManager;
import android.net.Uri;
import android.os.Build;
import android.util.Log;
import android.util.Base64;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.Drawable;

import androidx.annotation.Nullable;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DecodeFormat;
import com.bumptech.glide.load.model.GlideUrl;
import com.bumptech.glide.load.model.LazyHeaders;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Callback;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.views.imagehelper.ResourceDrawableIdHelper;
import com.facebook.react.views.imagehelper.ImageSource;

public class ManageWallpaperModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;
    private WallpaperManager wallpaperManager;
    private Callback rctCallback = null;
    private ReadableMap rctParams;
    private ResourceDrawableIdHelper mResourceDrawableIdHelper;
    private Uri mUri;
    private ReactApplicationContext mApplicationContext;
    private Activity mCurrentActivity;

    public ManageWallpaperModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        mApplicationContext = getReactApplicationContext();

        wallpaperManager = WallpaperManager.getInstance(mApplicationContext);
    }

    @Override
    public String getName() {
        return "ManageWallpaper";
    }

    public void sendMessage(String status, String msg, String url) {
        if (rctCallback != null) {
            WritableMap map = Arguments.createMap();
            map.putString("status", status);
            map.putString("msg", msg);
            map.putString("url", url);
            rctCallback.invoke(map);
            rctCallback = null;
        }
    }

    @ReactMethod
    public void setWallpaper(final ReadableMap params, final String type, Callback callback) {

        final String source = params.hasKey("uri") ? params.getString("uri") : null;
        ReadableMap headers = params.hasKey("headers") ? params.getMap("headers") : null;

        if (rctCallback != null) {
            WritableMap map = Arguments.createMap();

            map.putString("status", "error");
            map.putString("msg", "busy");
            map.putString("url", source);
            callback.invoke(map);
            return;
        }

        rctCallback = callback;
        rctParams = params;

        final CustomTarget<Bitmap> customTarget = this.getCustomTarget(source, type);
        mCurrentActivity = getCurrentActivity();

        if (mCurrentActivity == null) {
            sendMessage("error", "CurrentActivity is null", source);
        }

        //handle base64
        if ("data:image/png;base64,".startsWith(source)) {
            mCurrentActivity.runOnUiThread(new Runnable() {
                public void run() {
                    Util.assertMainThread();
                    try {
                        Glide
                            .with(mApplicationContext)
                            .asBitmap()
                            .apply(requestOptions())
                            .load(Base64.decode(source.replaceAll("data:image\\/.*;base64,", ""), Base64.DEFAULT))
                            .into(customTarget);
                    } catch (Exception e) {
                        sendMessage("error", "Exception in Glide：" + e.getMessage(), source);
                    }
                }
            });

            return;
        }

        boolean useStorageFile = false ;

        // handle bundled app resources
        try {
            mUri = Uri.parse(source);
            // Verify scheme is set, so that relative uri (used by static resources) are not handled.
            if (mUri.getScheme() == null) {
                mUri = null;
            } else if(
                !mUri.getScheme().equals("http") &&
                !mUri.getScheme().equals("https")
            ) {
                useStorageFile = true;
            }
        } catch (Exception e) {
            // ignore malformed uri, then attempt to extract resource ID.
        }

        if (mUri == null) {
            ImageSource is = new ImageSource(this.getReactApplicationContext(), source);
            if (is.isResource()) {
                int resId = mResourceDrawableIdHelper.getInstance().getResourceDrawableId(this.getReactApplicationContext(), source);
                Bitmap mBitmap = BitmapFactory.decodeResource(this.getReactApplicationContext().getResources(), resId);

                try {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        if (type.equals("home")) {
                            wallpaperManager.setBitmap(mBitmap, null, true, WallpaperManager.FLAG_SYSTEM);
                        } else if (type.equals("lock")) {
                            wallpaperManager.setBitmap(mBitmap, null, true, WallpaperManager.FLAG_LOCK);
                        } else {
                            wallpaperManager.setBitmap(mBitmap, null, true, WallpaperManager.FLAG_LOCK | WallpaperManager.FLAG_SYSTEM);   
                        }
                    } else {
                        wallpaperManager.setBitmap(mBitmap);
                    }
                    sendMessage("success", "Set Wallpaper Success", source);
                } catch (Exception e) {
                    sendMessage("error", "Exception in SimpleTarget：" + e.getMessage(), source);
                    return;
                }
                return;
            }

            mUri = is.getUri();
            mCurrentActivity.runOnUiThread(new Runnable() {
                public void run() {
                    Util.assertMainThread();
                    try{
                        Glide
                            .with(mApplicationContext)
                            .asBitmap()
                            .apply(requestOptions())
                            .load(mUri)
                            .into(customTarget);
                    } catch (Exception e) {
                        sendMessage("error", "Exception in Glide：" + e.getMessage(), source);
                    }
                }
            });
        } else if (useStorageFile) {
            mCurrentActivity.runOnUiThread(new Runnable() {
                public void run() {
                    Util.assertMainThread();
                    try {
                        Glide
                            .with(mApplicationContext)
                            .asBitmap()
                            .apply(requestOptions())
                            .load(mUri)
                            .into(customTarget);
                    } catch (Exception e) {
                        sendMessage("error", "Exception in Glide：" + e.getMessage(), source);
                    }
                }
            });
        } else {
            // Handle an http / https address
            final LazyHeaders.Builder lazyHeaders = new LazyHeaders.Builder();

            if (headers != null) {
                ReadableMapKeySetIterator it = headers.keySetIterator();
                Log.d("next headers", String.valueOf(it.hasNextKey()));
                while (it.hasNextKey()) {
                    String Key = it.nextKey();
                    lazyHeaders.addHeader(Key, headers.getString(Key));
                }
            }

            mCurrentActivity.runOnUiThread(new Runnable() {
                public void run() {
                    Util.assertMainThread();
                    try {
                        Glide
                            .with(mApplicationContext)
                            .asBitmap()
                            .apply(requestOptions())
                            .load(new GlideUrl(mUri.toString(), lazyHeaders.build()))
                            .into(customTarget);
                    } catch (Exception e) {
                        sendMessage("error", "Exception in Glide：" + e.getMessage(), source);
                    }
                }
            });
        }
    }

    private RequestOptions requestOptions() {
        return new RequestOptions()
                .format(DecodeFormat.PREFER_ARGB_8888)
                .centerCrop();
    }

    private CustomTarget<Bitmap> getCustomTarget(final String source, final String type) {
        return new CustomTarget<Bitmap>(1080, 1920) {
            @Override
            public void onResourceReady(Bitmap bitmap, Transition<? super Bitmap> transition) {
                try {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        if (type.equals("home")) {
                            wallpaperManager.setBitmap(bitmap, null, true, WallpaperManager.FLAG_SYSTEM);
                        } else if (type.equals("lock")) {
                            wallpaperManager.setBitmap(bitmap, null, true, WallpaperManager.FLAG_LOCK);
                        } else {
                            wallpaperManager.setBitmap(bitmap, null, true, WallpaperManager.FLAG_LOCK | WallpaperManager.FLAG_SYSTEM);   
                        }
                    } else {
                        wallpaperManager.setBitmap(bitmap);
                    }
                    sendMessage("success", "Set Wallpaper Success", source);
                } catch (Exception e) {
                    sendMessage("error", "Exception in SimpleTarget：" + e.getMessage(), source);
                    return;
                }
            }

            @Override
            public void onLoadCleared(@Nullable Drawable placeholder) {

            }

            @Override
            public void onLoadFailed(Drawable errorDrawable) {
                // Do nothing.
                sendMessage("error", "Set Wallpaper Failed：" + errorDrawable.toString(), source);
            }
        };
    }
}


example/App.js
/**
 * Sample React Native App
 *
 * adapted from App.js generated by the following command:
 *
 * react-native init example
 *
 * https://github.com/facebook/react-native
 */

import * as React from 'react';
import {Platform, View, TouchableOpacity, Text, StyleSheet} from 'react-native';
import {ManageWallpaper, TYPE} from 'react-native-manage-wallpaper';

export default class App extends React.Component<{}> {
  _setWallpaper = type => {
    const uri = {
      home:
        'https://i.pinimg.com/originals/76/5e/1d/765e1dc8cb1cc115fb3b0b39a895fdeb.jpg',
      lock: 'https://pbs.twimg.com/media/DLHyvP1VYAA4eCA.jpg',
      both: 'https://images.unsplash.com/photo-1473081556163-2a17de81fc97',
    };
    ManageWallpaper.setWallpaper(
      {
        uri: uri[type],
      },
      res => {
        console.warn('check', res);
      },
      type,
    );
  };

  render() {
    return (
      <View style={styles.container}>
        <TouchableOpacity
          style={styles.button}
          onPress={() => this._setWallpaper(TYPE.HOME)}>
          <Text style={styles.text}>Change Home Wallpaper</Text>
        </TouchableOpacity>
        {Platform.Version >= 24 && (
          <>
            <TouchableOpacity
              style={styles.button}
              onPress={() => this._setWallpaper(TYPE.LOCK)}>
              <Text style={styles.text}>Change Lock Wallpaper</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={styles.button}
              onPress={() => this._setWallpaper(TYPE.BOTH)}>
              <Text style={styles.text}>Change Both Wallpaper</Text>
            </TouchableOpacity>
          </>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  button: {
    paddingHorizontal: 30,
    paddingVertical: 8,
    marginBottom: 24,
    borderRadius: 16,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#000000',
  },
  text: {
    fontSize: 20,
    margin: 10,
    textAlign: 'center',
    color: '#FFFFFF',
  },
});
